### PR TITLE
Add 5.2.4 Operation Type Exists

### LINKED ISSUE #1097: No validation rule exists to assert mutation/subscription operations exist in the schema
When validating a document against a schema, there exists various validation rules for operations, but none to verify that if a mutation or subscription operation type is requested, that the schema is configured for such an operation.

For example, the below document would current pass all defined validation checks even with no mutation operation defined in the schema:

```gql
mutation {
  __typename
}
```

While per 6.2.2 the execution phase would fail the assertion that "mutationType is an Object type" and would fail to execute, this problem should be caught during validation, not execution.

Note that in the above sample, only 5.3.1 Field Selections could possibly fail validation.  For GraphQL.NET, each validation rule assumes that prior validation steps have passed.  For example, we do not raise errors about invalid fields when the parent field does not exist.  So no error is raised due to 5.3.1.  Further, `__typename` is a meta-field and would be not be compared to the schema in any case.  We perform the assertion at the execution step pursuant to the GraphQL spec.

I suggest adding validation rule 5.2.4 indicating that for mutation or subscription operation types, the schema must be configured for the requested operation.

### PR DESCRIPTION
**Pull request by Shane32**
Closes #1097

### CODE REVIEW

File: spec/Section 5 -- Validation.md [hunk 1]
```diff
@@ -332,6 +332,37 @@ contain any number of operations, each of which may contain different root
 fields. When executed, a document containing multiple subscription operations
 must provide the operation name as described in {GetOperation()}.
 
+### 5.2.4 Operation Type Configuration
```
Comment (Shane32) [inline spec/Section 5 -- Validation.md:335]:
```suggestion
### Operation Type Configuration
```

File: spec/Section 5 -- Validation.md [hunk 2]
```diff
@@ -332,6 +332,37 @@ contain any number of operations, each of which may contain different root
 fields. When executed, a document containing multiple subscription operations
 must provide the operation name as described in {GetOperation()}.
 
+### Operation Type Configuration
+
+#### Formal Specification
+
+For each operation definition in the document:
+- Let operationType be the type of the operation (query, mutation, or subscription).
+- The corresponding type for operationType must be defined in the schema.
```
Comment (Shane32) [inline spec/Section 5 -- Validation.md:341]:
```suggestion
- Let {operationType} be the type of the operation (query, mutation, or subscription).
- The corresponding type for {operationType} must be defined in the schema.
```

File: spec/Section 5 -- Validation.md [hunk 3]
```diff
@@ -332,6 +332,37 @@ contain any number of operations, each of which may contain different root
 fields. When executed, a document containing multiple subscription operations
 must provide the operation name as described in {GetOperation()}.
 
+### Operation Type Configuration
+
+#### Formal Specification
+
+For each operation definition in the document:
```
Comment (Shane32) [inline spec/Section 5 -- Validation.md:339]:
```suggestion
For each operation definition in the document:

```

**Review by benjie — DISMISSED** [review]
Great find!

**Review by benjie — APPROVED** [review]
Looks good to me, though some slight editorial on the wording to match the style of the spec as a whole might be warranted.

**Review by JoviDeCroock — APPROVED** [review]
This is a great validation rule 🙌

**Comment by benjie** [comment]
The lack of this validation rule can be demonstrated in GraphQL.js with this:

```ts
import { GraphQLSchema, GraphQLInt, graphql, GraphQLObjectType } from 'graphql';

const Query = new GraphQLObjectType({
  name: 'Query',
  fields: {
    a: {
      type: GraphQLInt,
      resolve() {
        return 42;
      },
    },
  },
});

const schema = new GraphQLSchema({
  query: Query,
});

const result = await graphql({ schema, source: `mutation { __typename }` });
console.dir(result);
```

which produces:

```
{
  data: null,
  errors: [
    GraphQLError: Schema is not configured to execute mutation operation.
        at executeOperation [...]
      path: undefined,
      locations: [Array],
      extensions: [Object: null prototype] {}
    }
  ]
}
```

We should really mark this as an invalid request, and thus it should not have a `data` key in the response.

**Comment by yaacovCR** [comment]
This is a refresh of #955 which has a pending PR at the implementation https://github.com/graphql/graphql-js/pull/3601

**Comment by benjie** [comment]
Good spot @yaacovCR; it seems both @Shane32 and I had commented on that previous PR!

I have merged this PR into that PR and applied my own editorial edits on top of both:

- https://github.com/graphql/graphql-spec/pull/955

**Pull request closed without merging**